modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/ApplicativeError.kt

```diff
@@ -36,7 +43,11 @@ interface ApplicativeError<F, E> : Applicative<F> {
     try {
       just(f())
     } catch (t: Throwable) {
-      raiseError<A>(recover(t))
+      if (NonFatal(t)) {
```

Comment:
We can encapsulate this pattern For All `ApplicativeError<F>` which constitutes most of the changes in this PR.
If you place this function as part of the MonadThrow<F> type class
```kotlin
fun <F, A> Throwable.raiseNonFatal(): Kind<F, A> =
  if (NonFatal(this)) raiseError<A>(this) else throw e
```

This functions is scoped to the MonadThrow receiver and will be available in all instances of all the data types as extension functions and would allow you to remove the if else boilerplate in most cases below.

Reply:
Thanks for the review and the great suggestion. :) I've applied it to all but IOBracket, it didn't work there (I guess because it's not running in the MonadThrow context?)

There is also a AbstractMethodError in `IOTest MonadDefer bind: binding failure` that I can't quite figure out.

Reply:
Change the genError code to throw something else instead, maybe that's the issue. I'd add a new generator that throws one of the fatal errors and test for that.

Reply:
After updating graalvm to rc12 (my jdk), the test no longer fails.
Should I add the fatal error test to MonadDefer laws?

Reply:
I'd add a MonadThrow laws altogether that have all MonadErrorLaws, put that test in them, and make MonadDefer do MonadThrowLaws instead of MonadErrorLaws.

That way we have tests where the functions are defined in the hierarchy :D

Reply:
Mmh, does this test even work inside MonadThrow? There is no delay/defer that actually does a try {} catch {} or am I missing something?

Reply:
Then the function should live in MonadDefer :D @raulraja opinions?

Reply:
The function can be expressed in `MonadThrow` and override `catch` to use it:
```kotlin
override fun <A> catch(recover: (Throwable) -> Throwable, f: () -> A): Kind<F, A> 
override fun <A> ApplicativeError<F, Throwable>.catch(f: () -> A): Kind<F, A> 
```
It has application outside suspensions of effects given the loose approach we take with `Try` at the moment. This is an example of a datatype that would benefit from it and that does not allow suspension and can't implement MonadDefer. Another one would be Either for partially applied Throwables.
IMO it should live in `MonadThrow`.

Reply:
Sorry, I'm kind of lost now. :D I've added the suggested function `raiseNonFatal` to MonadThrow.
What else can I do? (and how can I/should I test it as MonadThrow Law?)

Reply:
@kartoffelsup this should cover it:

- override the two functions above using `raiseNonFatal` to implement them in `MonadThrow`
- Add a new law in the `arrow.test.laws.MonadErrorLaws` 
> Monad Error Laws: NonFatal is caught
- Add a new la  in the `arrow.test.laws.MonadErrorLaws` 
> Monad Error Laws: Fatal errors are thrown
- Add kdoc and a `kotlin:ank:playground` snippet similar to the one in `Functor` that illustrates the usage of NonFatal in the `NonFatal` object.

That should do it.

Reply:
@raulraja thanks for the guidance!
Regarding the catch overrides, atm I've already overridden them in ApplicativeError like so:
```kotlin
 fun <A> catch(recover: (Throwable) -> E, f: () -> A): Kind<F, A> =
    try {
      just(f())
    } catch (t: Throwable) {
      if (NonFatal(t)) {
        raiseError<A>(recover(t))
      } else {
        throw t
      }
    }

  fun <A> ApplicativeError<F, Throwable>.catch(f: () -> A): Kind<F, A> =
    catch(::identity, f)
```
Do we really need the override in MonadThrow? 
Also, it looks to me that the `catch(recover: (Throwable) -> E, f: () -> A)` variant can't be done with `raiseNonFatal` as I am unable to call `#map` on it for the `recover` function.

Reply:
no, if you already did it there it's fine how you did it.

Reply:
I've applied the changes you requested. Please let me know if there is anything else I can do. Thanks again for the help!